src/rules/ban_types.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 the Deno authors. All rights reserved. MIT license.
+use super::Context;
+use super::LintRule;
+use swc_atoms::JsWord;
+use swc_ecma_visit::Node;
+use swc_ecma_visit::Visit;
+
+pub struct BanTypes;
+
+impl LintRule for BanTypes {
+  fn new() -> Box<Self> {
+    Box::new(BanTypes)
+  }
+
+  fn code(&self) -> &'static str {
+    "ban-types"
+  }
+
+  fn lint_module(&self, context: Context, module: swc_ecma_ast::Module) {
+    let mut visitor = BanTypesVisitor::new(context);
+    visitor.visit_module(&module, &module);
+  }
+}
+
+struct BanTypesVisitor {
+  context: Context,
+}
+
+impl BanTypesVisitor {
+  pub fn new(context: Context) -> Self {
+    Self { context }
+  }
+}
+
+const BANNED_TYPES: [(&str, &str); 6] = [
+  ("String", "Use `string` instead"),
+  ("Boolean", "Use `boolean` instead"),
+  ("Number", "Use `number` instead"),
+  ("Symbol", "Use `symbol` instead"),
+  ("Function", "Define the function shape Explicitly."),
+  ("Object", "if you want a type meaning `any object` use `Record<string, unknown>` instead,
+or If you want a type meaning `any value`, you probably want `unknown` instead."),
+];
+
+impl Visit for BanTypesVisitor {
+  fn visit_ts_type_ref(
+    &mut self,
+    ts_type_ref: &swc_ecma_ast::TsTypeRef,
+    _parent: &dyn Node,
+  ) {
+    if let swc_ecma_ast::TsEntityName::Ident(ident) = &ts_type_ref.type_name {
+      if let Some((_, message)) = BANNED_TYPES
+        .iter()
+        .find(|banned_type| JsWord::from(banned_type.0) == ident.sym)
+      {
+        self
+          .context
+          .add_diagnostic(ts_type_ref.span, "ban-types", message);
+      }
+    }
+    if let Some(type_param) = &ts_type_ref.type_params {
+      self.visit_ts_type_param_instantiation(type_param, ts_type_ref);
+    }
+  }
+  fn visit_ts_type_lit(
+    &mut self,
+    ts_type_lit: &swc_ecma_ast::TsTypeLit,
+    _parent: &dyn Node,
+  ) {
+    if !ts_type_lit.members.is_empty() {
+      for element in ts_type_lit.members.iter() {
+        self.visit_ts_type_element(element, ts_type_lit);
+      }
+      return;
+    }
+    self.context.add_diagnostic(
+        ts_type_lit.span,
+        "ban-types",
+        "if you want a type meaning `any object` use `Record<string, unknown>` instead,
+or If you want a type meaning `any value`, you probably want `unknown` instead.",
```

Comment:
Ditto, also, already defined in `BANNED_TYPES` array - use that instead?

Reply:
of course, thanks